janusgraph-test/src/main/java/org/janusgraph/graphdb/JanusGraphTest.java

```diff
@@ -3333,6 +3332,279 @@ public void testRelationTypeIndexes() {
+        JanusGraphVertex v2 = tx.addVertex("user");
+        v1.addEdge("knows", v2, "id", 10);
+        clopen(option(AUTO_TYPE), "none");
+
+        v1 = tx.addVertex("user");
+        v2 = tx.addVertex("user");
+        v1.addEdge("knows", v2, "id", 10);
+
+        try {
+            v1.addEdge("knows", v2, "test", 10);
+            fail("This should never reached!");
+        } catch (IllegalArgumentException ignored) {
+        }
+    }
+
+    private void createStrictSchemaForConnections() {
+        clopen(option(AUTO_TYPE), "none");
+        VertexLabel user = mgmt.makeVertexLabel("user").make();
+        VertexLabel company = mgmt.makeVertexLabel("company").make();
+        EdgeLabel edge = mgmt.makeEdgeLabel("knows").make();
+        mgmt.makeEdgeLabel("buys").make();
+        mgmt.addConnection(edge, user, company);
+        finishSchema();
+    }
+
+    @Test
+    public void testEnforcedSchemaAllowsDefinedConnections() {
+        createStrictSchemaForConnections();
+
+        JanusGraphVertex v1 = tx.addVertex("user");
+        JanusGraphVertex v2 = tx.addVertex("company");
+        v1.addEdge("knows", v2);
+    }
+
+    @Test
+    public void testSchemaIsEnforcedForConnections() {
+        createStrictSchemaForConnections();
+
+        JanusGraphVertex v1 = tx.addVertex("user");
+        try {
+            v1.addEdge("buys", v1);
+            fail("This should never reached!");
+        } catch (IllegalArgumentException ignored) {
+        }
+
+        JanusGraphVertex v2 = tx.addVertex("company");
+        try {
+            v2.addEdge("knows", v1);
+            fail("This should never reached!");
+        } catch (IllegalArgumentException ignored) {
+        }
+    }
+
+    @Test
+    public void testAutoSchemaMakerForConnectionConstraints() {
+        JanusGraphVertex v1 = tx.addVertex("user");
+        JanusGraphVertex v2 = tx.addVertex("user");
+        v1.addEdge("knows", v2);
+        clopen(option(AUTO_TYPE), "none");
+
+        v1 = tx.addVertex("user");
+        v2 = tx.addVertex("user");
+        v1.addEdge("knows", v2);
+
+        try {
+            v1.addEdge("has", v2);
+            fail("This should never reached!");
+        } catch (IllegalArgumentException ignored) {
+        }
+    }
+
+    @Test
+    public void testSupportChangeNameOfEdgeAndUpdateConnections() {
+        clopen(option(AUTO_TYPE), "none");
+        VertexLabel user = mgmt.makeVertexLabel("user").make();
+        VertexLabel company = mgmt.makeVertexLabel("company").make();
+        EdgeLabel edge = mgmt.makeEdgeLabel("knows").make();
+        mgmt.makeEdgeLabel("buys").make();
+        mgmt.addConnection(edge, user, company);
+        finishSchema();
+
+        JanusGraphVertex v1 = tx.addVertex("user");
+        JanusGraphVertex v2 = tx.addVertex("company");
+        v1.addEdge("knows", v2);
+        newTx();
+
+        edge = mgmt.getEdgeLabel("knows");
+        mgmt.changeName(edge, "know");
```

Comment:
Changing the name from _knows_ to _know_ doesn't make it easy to spot the difference. I think that it would be better if those two names would differ in more than one character.